core/src/main/scala/com/databricks/labs/remorph/preprocessor/README.md

```diff
@@ -0,0 +1,205 @@
+     template `{% for payment_method in payment_methods -%}` is located in the middle of a SQL statement. In this case
+     we would need to allow templates to occur anywhere in the parse in violation of the normal syntax rules. That is essentionally, not viable.
+ - Jinga allows line statements, also with the prefix being configurable. Hence, we need to be able to handle
+   them too. Typically, they would start with a single prefix such as `# for item in seq`, and the entire line
+   is a ninja statement.
+ - There is nothing stopping a user from stuffing variables with actual SQL statements. We will probably draw a
+   line at supporting that.
+ - Macros can contain bits of SQL code, which we may be able to attempt to translate by trying different entry points
+   into the SQL parser. However, this is not guaranteed to be successful as literalyl anything can be put anywhere 
+   in macros
+ - There are macros that do things like add or not add commas at the ends for loops of text generation. We can cater
+   for this using the generally accepted practice of allow trailing commas in sequences, so we can accept partial
+   lists of columns, for example.
+
+## Approaches to handling templates
+
+We should first note that our template handler/preprocessor is not expanding the templates, but locating them and
+translating embedded SQL. The templates themselves will be left in place, and the translated SQL will be adorned
+with them. We are writing a XXX SQL plus DBT to Databricks SQL plus DBT, such that users will then maintain the
+Databricks SQL alongside their templates.
+
+We therefore have the following approaches:
+
+### Lexing and Placeholding
+
+As there are few constructs in the templates, we can lex the raw input, store the gathered tokens, and replace
+them with a single token identifying the template. The TSQL grammars then need only look for simple template
+references and not have to worry about the template syntax itself. The pre-processor here becomes more of a controller
+in that it can transform pieces of the input text by calling the source dialect transpiler and the code generator
+for each component of the source text. However, the grammars for the source dialects will need to be modified
+to accept the template tokens at strategic points in the parse.
+
+#### Advantages
+
+ - The lexer and parser are not burdened with the template syntax
+ - The lexer and parser can be used with a few modifications
+
+#### Disadvantages
+
+ - Templates can be placed anywhere, so the grammar will need to be strategically adorned with valid
+   places that templates can occur. This is not a trivial task, however it is doable if we accept that
+   we will not be able to process some esoteric uses of templates, which are generally abuses of the concept.
+   For instance `SELECT A {{ , }} B {{ , }} C` is not a reasonable construct. It would need to be dealt
```

Comment:
leave it as the `UnresolvedNode` and we'll add a support for it if it's used often